## Sentiment Analysis for Emotion Classification

### 1. Introduction

#### Sentiment Analysis is a Natural Language Processing (NLP) task used to identify emotions or opinions in text. In this assignment, we build machine learning models to classify emotions using text data.


#### 2. Loading and Preprocessing
##### 2.1 Import Required Libraries

In [12]:
import pandas as pd
import numpy as np
import re
import nltk
import matplotlib.pyplot as plt

from nltk.corpus import stopwords
from nltk.tokenize import word_tokenize

nltk.download('punkt')
nltk.download('punkt_tab')
nltk.download('stopwords')

[nltk_data] Downloading package punkt to
[nltk_data]     C:\Users\vanda\AppData\Roaming\nltk_data...
[nltk_data]   Package punkt is already up-to-date!
[nltk_data] Downloading package punkt_tab to
[nltk_data]     C:\Users\vanda\AppData\Roaming\nltk_data...
[nltk_data]   Unzipping tokenizers\punkt_tab.zip.
[nltk_data] Downloading package stopwords to
[nltk_data]     C:\Users\vanda\AppData\Roaming\nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


True

##### 2.2 Load Dataset

In [13]:
df = pd.read_csv('nlp_dataset.csv')  
df.head()

,Comment,Emotion
0,i seriously hate one subject to death but now ...,fear
1,im so full of life i feel appalled,anger
2,i sit here to write i start to dig out my feel...,fear
3,ive been really angry with r and i feel like a...,joy
4,i feel suspicious if there is no one outside l...,fear


##### 2.3 Text Preprocessing
##### Steps Performed
- Convert text to lowercase
- Remove special characters
- Tokenization
- Remove stopwords

In [14]:
# Clean column names 
df.columns = df.columns.str.strip().str.lower()

# Define stopwords
stop_words = set(stopwords.words('english'))

def preprocess_text(text):
    # Handle missing values
    if pd.isna(text):
        return ""
    
    # Convert to string
    text = str(text)
    
    # Lowercase
    text = text.lower()
    
    # Remove special characters
    text = re.sub(r'[^a-zA-Z\s]', '', text)
    
    # Tokenization
    tokens = word_tokenize(text)
    
    # Remove stopwords
    tokens = [word for word in tokens if word not in stop_words]
    
    return " ".join(tokens)

df['clean_text'] = df['comment'].apply(preprocess_text)
y = df['emotion']

##### 2.4 Explanation of Preprocessing
- Lowercasing → Ensures uniformity
- Removing special characters → Eliminates noise
- Tokenization → Splits text into words
- Stopword removal → Removes common words (e.g., “the”, “is”) that don’t add meaning
##### These steps improve model performance by reducing noise and focusing on meaningful words.

#### 3. Feature Extraction
##### 3.1 Using TF-IDF Vectorizer

In [15]:
from sklearn.feature_extraction.text import TfidfVectorizer

vectorizer = TfidfVectorizer(max_features=5000)

X = vectorizer.fit_transform(df['clean_text'])
y = df['emotion']   

##### 3.2 Explanation
- TF-IDF (Term Frequency–Inverse Document Frequency) assigns importance to words
- Words that appear frequently in one document but not across all documents get higher weight
- Converts text into numerical vectors usable by ML models

#### 4. Model Development
##### 4.1 Train-Test Split

In [16]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42
)

#### 4.2 Naive Bayes Model

In [17]:
from sklearn.naive_bayes import MultinomialNB
from sklearn.metrics import classification_report

nb_model = MultinomialNB()
nb_model.fit(X_train, y_train)

y_pred_nb = nb_model.predict(X_test)

print("Naive Bayes Performance:\n")
print(classification_report(y_test, y_pred_nb))

Naive Bayes Performance:

              precision    recall  f1-score   support

       anger       0.88      0.95      0.91       392
        fear       0.91      0.92      0.92       416
         joy       0.94      0.86      0.90       380

    accuracy                           0.91      1188
   macro avg       0.91      0.91      0.91      1188
weighted avg       0.91      0.91      0.91      1188



#### 4.3 Support Vector Machine (SVM)

In [18]:
from sklearn.svm import SVC

svm_model = SVC()
svm_model.fit(X_train, y_train)

y_pred_svm = svm_model.predict(X_test)

print("SVM Performance:\n")
print(classification_report(y_test, y_pred_svm))

SVM Performance:

              precision    recall  f1-score   support

       anger       0.93      0.94      0.94       392
        fear       0.97      0.90      0.94       416
         joy       0.90      0.97      0.94       380

    accuracy                           0.94      1188
   macro avg       0.94      0.94      0.94      1188
weighted avg       0.94      0.94      0.94      1188



#### 5. Model Comparison

##### 5.1 Evaluation Metrics

In [19]:
from sklearn.metrics import accuracy_score, f1_score

print("Naive Bayes Accuracy:", accuracy_score(y_test, y_pred_nb))
print("SVM Accuracy:", accuracy_score(y_test, y_pred_svm))

print("Naive Bayes F1:", f1_score(y_test, y_pred_nb, average='weighted'))
print("SVM F1:", f1_score(y_test, y_pred_svm, average='weighted'))

Naive Bayes Accuracy: 0.9090909090909091
SVM Accuracy: 0.9351851851851852
Naive Bayes F1: 0.9089193266783082
SVM F1: 0.9351828096425033


### Model Comparison

##### The performance of the models was evaluated using:

- Accuracy → Measures overall correctness of predictions
- F1-score → Balances precision and recall, especially useful for multi-class classification
##### Naive Bayes
- Achieved an accuracy of 90.91%
- Performs well for text classification tasks
- Faster and computationally efficient
- However, it assumes independence between features, which may limit performance
##### Support Vector Machine (SVM)
- Achieved a higher accuracy of 93.52%
- Produced a better F1-score (0.9352)
- Handles high-dimensional TF-IDF features effectively
- Captures complex relationships between words better than Naive Bayes

#### Based on the results, <span style="color:red">Support Vector Machine (SVM)</span> out performs Naive Bayes.

##### Reason:

- Higher Accuracy (+2.6% improvement)
- Better F1-score
- More effective in handling complex text patterns
- SVM performed better due to its ability to handle high-dimensional sparse data like TF-IDF.

### Conclusion

- In this assignment, machine learning models were developed to classify emotions from textual data. The dataset was preprocessed through steps such as lowercasing, removal of special characters, tokenization, and stopword removal. These preprocessing techniques helped in improving data quality by reducing noise and retaining meaningful information.TF-IDF vectorization was used to transform the cleaned text into numerical features, enabling the models to process textual data effectively. Two models, Naive Bayes and Support Vector Machine (SVM), were trained and evaluated.The results show that both models performed well, with Naive Bayes achieving over 90% accuracy. However, SVM provided superior performance, achieving 93.52% accuracy and the highest F1-score, making it the better model for this task. This is because SVM is more capable of handling high-dimensional data and capturing complex relationships in text.Overall, this assignment highlights the importance of preprocessing and feature extraction in NLP tasks, and demonstrates that advanced models like SVM can significantly improve performance in emotion classification problems.